Added extents for GPU and CPU Particles2D emission points

eligt opened the pull request:
The emission mask now uses relative values which are multiplied by the emission extents. This also fixes the issue where emission points are dependent on original texture size.

This commits also change emission points texture to be centered on the particle system instead of placing the top left corner at the center.

I think this also fixes #16411

clayjohn commented:
Is this change relevant for CPUParticles2D as well? If so, that needs to be implemented as well.

eligt committed:
Added extents for GPU particles2d emission points

The emission mask now uses relative values which are multiplied by the emission extents. This also fixes the issue where emission points are dependent on original texture size.

This commits also change emission points texture to be centered on the particle system instead of placing the top left corner at the center.

I think this also fixes #16411
Added CPU particles

eligt commented:
@clayjohn Done.

aaronfranke commented:
@eligt Is this still desired? If so, it needs to be rebased on the latest master branch.

Lambdanaut commented:
@aaronfranke @eligt I'd still really like to see this change.

QbieShay commented:
I'm not convinced that this is the right solution for the problem at hand.
There are two issues being addressed here:
- the points are bound to the texture size, and they shouldn't be
- the point center is on 0,0 instead of half the extents
I think that offsetting by half the size is fine, but what I wouldn't do is the rescaling in the way that is done here: the implementation in this PR makes it necessary to specify the extents of the box, so there is an additional setup stage, where things won't work out of the box. I'd substitute the usage of box extents with a single float scale property that defaults to one and remove the normalization from the creation of the points texture, so that it can be scaled relatively to the size of the initial texture and not with an absolute value.

YuriSizov commented:
Seems like there is no consensus on the solution, nor has there been any activity with this PR for a while. Since we're at the end of the beta stage, I have to close this. An agreed upon solution to the presented problems that doesn't break compatibility is likely still welcome in future.
